crates/hir/src/has_source.rs

```diff
@@ -146,7 +160,30 @@ impl HasSource for Variant {
 impl HasSource for Function {
     type Ast = ast::Fn;
     fn source(self, db: &dyn HirDatabase) -> Option<InFile<Self::Ast>> {
-        Some(self.id.lookup(db).source(db))
+        match self.id {
+            AnyFunctionId::FunctionId(id) => Some(id.loc(db).source(db)),
+            // HACK: When calling `source()`, we use the trait method source, but when calling `source_with_range()`,
```

Comment:
I don't think we should label this a hack, we will likely end up with more non-`source` things in the future (think theoretical rlib support)